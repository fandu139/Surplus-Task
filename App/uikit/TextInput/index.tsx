import React from 'react';
import AppStyles from '../../theme/appStyles';
import Colors from '../../theme/colors';
import Fonts from '../../theme/fonts';
import Icon from '../Icon';
import Pattern from '../../helper/pattern';
import { TextField } from 'rn-material-ui-textfield';
import { View, StyleSheet, StyleProp, ViewStyle, TextStyle, Text } from 'react-native';

/**
 * Material text input
 * @param props
 * @param label
 * @param text
 * @param onChangeText
 * @param disabled
 * @param isError
 * @param errorText
 * @param style
 * @returns {*}
 * @constructor
 */

type RightIcon =
  | {
      size: number;
      name: string;
      color: string;
      onPress: () => void;
      testID?: string;
    }
  | undefined;

type KeyboardType =
  | 'default'
  | 'number-pad'
  | 'decimal-pad'
  | 'numeric'
  | 'email-address'
  | 'phone-pad'
  | 'ascii-capable'
  | 'numbers-and-punctuation'
  | 'url'
  | 'name-phone-pad'
  | 'twitter'
  | 'web-search'
  | 'visible-password'
  | string;

type CustomColor = {
  selectionColor: string;
  base: string;
  tintColor: string;
  textColor: string;
  labelColor: string;
};

type TextAlign = 'left' | 'center' | 'right';

interface Props {
  label: string;
  text?: string;
  onChangeText: (value: string) => void;
  disabled?: boolean;
  isError?: boolean;
  errorText?: string;
  style?: StyleProp<TextStyle>;
  inverted?: boolean;
  autoCompleteType?:
    | 'cc-csc'
    | 'cc-exp'
    | 'cc-exp-month'
    | 'cc-exp-year'
    | 'cc-number'
    | 'email'
    | 'name'
    | 'password'
    | 'postal-code'
    | 'street-address'
    | 'tel'
    | 'username'
    | 'off';
  secureTextEntry?: boolean;
  rightIcon?: RightIcon;
  bottomRightText?: string;
  onPressRightText?: () => void;
  placeholder?: string;
  containerStyle?: StyleProp<ViewStyle>;
  keyboardType?: KeyboardType;
  onBlur?: (value: string) => void;
  bottomLabelStyle?: StyleProp<TextStyle>;
  fontSize?: number;
  contentInset?: Record<string, unknown>;
  lineContainer?: StyleProp<ViewStyle>;
  outerPrefix?: React.ReactNode;
  suffix?: React.ReactNode;
  customColor?: CustomColor;
  multiline?: boolean;
  numberOfLines?: number;
  textAlign?: TextAlign;
  maxLength?: number;
  testID: string;
  testIDHelper?: string;
  testIDRightText?: string;
  accessibilityLabel?: string;
  accessibilityLabelHelper?: string;
  accessibilityLabelRightText?: string;
  required?: boolean;
}

type TextInputComponent = React.ForwardRefExoticComponent<Props & React.RefAttributes<TextField>>;

const TextInput: TextInputComponent = React.forwardRef<TextField, Props>(
  (
    {
      label,
      text,
      onChangeText,
      disabled,
      isError,
      errorText,
      style,
      inverted,
      autoCompleteType,
      secureTextEntry,
      rightIcon,
      bottomRightText,
      onPressRightText,
      placeholder,
      containerStyle,
      keyboardType,
      onBlur,
      bottomLabelStyle,
      fontSize,
      contentInset,
      lineContainer,
      outerPrefix,
      suffix,
      customColor,
      multiline,
      numberOfLines,
      textAlign,
      maxLength,
      testID,
      testIDHelper,
      testIDRightText,
      accessibilityLabel,
      accessibilityLabelHelper,
      accessibilityLabelRightText,
      required,
    }: Props,
    ref,
  ) => {
    let color = {
      selectionColor: Colors.GRAY,
      base: text ? Colors.BLACK : Colors.GRAY80,
      tintColor: Colors.GRAY80,
      textColor: Colors.BLACK,
      labelColor: Colors.GRAY80,
    };

    if (inverted) {
      color = {
        selectionColor: Colors.WHITE,
        base: Colors.WHITE,
        tintColor: Colors.WHITE,
        textColor: Colors.WHITE,
        labelColor: Colors.WHITE,
      };
    }

    if (customColor) {
      color = {
        selectionColor: customColor.selectionColor,
        base: customColor.base,
        tintColor: customColor.tintColor,
        textColor: customColor.textColor,
        labelColor: customColor.labelColor,
      };
    }

    if (text) {
      const regex = Pattern.unicode.test(text);

      if (!regex) {
        isError = true;
        errorText = "Karakter yang Anda masukkan tidak valid";
      }
    }

    const textLabel = required ? undefined : label;

    return (
      <View style={containerStyle}>
        {required && (
          <View style={AppStyles.rowItemsCenter}>
            <Text style={styles.label}>{label}</Text>
            <Text testID={`${testID}-required`} style={styles.labelRequired}>
              {' '}
              *
            </Text>
          </View>
        )}
        <TextField
          ref={ref}
          allowFontScaling={false}
          textAlign={textAlign}
          multiline={multiline}
          numberOfLines={numberOfLines}
          bottomLabelStyle={bottomLabelStyle}
          onBlur={onBlur}
          keyboardType={keyboardType}
          useNativeDriver
          placeholder={placeholder}
          selectionColor={color.selectionColor}
          labelTextStyle={styles.labelTextStyle}
          titleTextStyle={styles.textStyle}
          secureTextEntry={secureTextEntry}
          autoCompleteType={autoCompleteType}
          underlineColor={Colors.GRAY}
          labelFontSize={Fonts.size.extraSmall}
          disabledLineType="solid"
          fontSize={fontSize}
          maxLength={maxLength}
          style={[styles.textInputStyle, style]}
          label={textLabel}
          value={text}
          onChangeText={onChangeText}
          disabled={disabled}
          lineWidth={1}
          activeLineWidth={1}
          disabledLineWidth={1}
          baseColor={color.base}
          tintColor={color.tintColor}
          textColor={color.textColor}
          labelColor={color.labelColor}
          errorColor={Colors.RED}
          lineColor={Colors.GRAYE0}
          disabledLineColor={Colors.GRAYE0}
          lineContainer={lineContainer}
          outerPrefix={outerPrefix}
          contentInset={{
            top: 0,
            label: 6,
            input: 10,
            ...contentInset,
          }}
          error={isError && errorText}
          bottomRightText={bottomRightText}
          onPressRightText={onPressRightText}
          suffix={
            rightIcon ? (
              <Icon
                style={styles.rightIconBase}
                size={rightIcon.size}
                name={rightIcon.name}
                color={rightIcon.color}
                onPress={rightIcon.onPress}
                testID={rightIcon.testID || 'right-icon-text-input'}
              />
            ) : (
              suffix
            )
          }
          text={text}
          testID={testID}
          testIDHelper={testIDHelper}
          testIDRightText={testIDRightText}
          accessibilityLabel={accessibilityLabel || testID}
          accessibilityLabelHelper={accessibilityLabelHelper || testIDHelper}
          accessibilityLabelRightText={accessibilityLabelRightText || testIDRightText}
        />
      </View>
    );
  },
);

const styles = StyleSheet.create({
  textInputStyle: {
    ...Fonts.regular,
    fontSize: Fonts.size.medium,
  },
  textStyle: {
    ...Fonts.regular,
    color: Colors.WHITE,
  },
  labelTextStyle: {
    ...Fonts.regular,
    fontSize: Fonts.size.extraSmall,
    color: Colors.GRAY80,
  },
  rightIconBase: {
    padding: 5,
  },
  label: {
    fontSize: Fonts.size.extraSmall,
    color: Colors.GRAY80,
  },
  labelRequired: {
    fontSize: Fonts.size.extraSmall,
    color: Colors.RED,
  },
});

TextInput.defaultProps = {
  text: undefined,
  disabled: false,
  isError: false,
  errorText: '',
  style: {},
  inverted: false,
  autoCompleteType: undefined,
  secureTextEntry: false,
  rightIcon: undefined,
  bottomRightText: undefined,
  onPressRightText: undefined,
  placeholder: undefined,
  containerStyle: {},
  keyboardType: undefined,
  onBlur: undefined,
  bottomLabelStyle: {},
  fontSize: Fonts.size.medium,
  contentInset: {},
  lineContainer: {},
  outerPrefix: undefined,
  suffix: undefined,
  customColor: undefined,
  multiline: false,
  numberOfLines: undefined,
  textAlign: undefined,
  maxLength: 255,
  testIDHelper: 'text-input-default-helper',
  testIDRightText: 'text-input-default-right-text',
  accessibilityLabel: undefined,
  accessibilityLabelHelper: undefined,
  accessibilityLabelRightText: undefined,
  required: false,
};

const TextInputGrouped = Object.assign(TextInput);

export default TextInputGrouped;
