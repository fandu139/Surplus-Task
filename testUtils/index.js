import AuthContext from '../App/context/Auth';
import React from 'react';
import { act, fireEvent, render } from '@testing-library/react-native';

const customRender = (ui, options) => {
  const isAuthenticated = options?.isAuthenticated || false;
  const renderOptions = options?.rendererOptions;
  const authContext = options?.authContext;
  return render(
    <AuthContext.Provider value={{ isAuthenticated, ...authContext }}>{ui}</AuthContext.Provider>,
    renderOptions,
  );
};

// override render method
export { customRender as render };

export function wait(ms) {
  return act(() => new Promise((r) => setTimeout(r, ms)));
}

// Due to debounce on Button UI Kit, we need to add milisecond waiting timer
export function fireAsyncPress(element, ms = 100) {
  fireEvent.press(element);
  return wait(ms);
}
