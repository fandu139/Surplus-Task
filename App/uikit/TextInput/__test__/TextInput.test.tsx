import TextInput from '../';
import React from 'react';
import { render } from '../../../../testUtils';

describe('TextInput component test', () => {
  describe('TextInput snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <TextInput
          testID="text-input-default"
          label="Label TextInput"
          onChangeText={() => undefined}
          text="Value TextInput"
        />,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('TextInput inverted snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <TextInput
          testID="text-input-default"
          label="Label TextInput"
          onChangeText={() => undefined}
          text="Value TextInput"
          inverted
        />,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('TextInput with custom color snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <TextInput
          testID="text-input-default"
          label="Label TextInput"
          onChangeText={() => undefined}
          text="Value TextInput"
          customColor={{
            selectionColor: 'gray',
            base: 'white',
            tintColor: 'blue',
            textColor: 'black',
            labelColor: 'black',
          }}
        />,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
