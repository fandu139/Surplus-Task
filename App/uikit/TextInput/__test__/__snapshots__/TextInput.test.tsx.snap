// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput component test TextInput inverted snapshot test should render correctly 1`] = `
<View
  style={{}}
>
  <View
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="auto"
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "height": 52,
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "borderColor": "rgba(255, 255, 255, 1)",
              "borderStyle": "solid",
              "borderWidth": 1,
              "bottom": 0,
              "left": -1.5,
              "position": "absolute",
              "right": -1.5,
              "top": -1.5,
            }
          }
        />
      </View>
      <View
        pointerEvents="auto"
        style={
          {
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "left": "-100%",
              "paddingLeft": "50%",
              "position": "absolute",
              "top": 0,
              "transform": [
                {
                  "scale": 0.75,
                },
                {
                  "translateY": 0,
                },
                {
                  "translateX": 0,
                },
              ],
              "width": "200%",
            }
          }
        >
          <Text
            collapsable={false}
            numberOfLines={1}
            style={
              {
                "color": "rgba(255, 255, 255, 1)",
                "fontSize": 16,
                "fontWeight": "normal",
                "includeFontPadding": false,
                "lineHeight": 16,
                "textAlign": "left",
                "textAlignVertical": "top",
              }
            }
          >
            Label TextInput
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            accessibilityLabel="text-input-default"
            allowFontScaling={false}
            autoCapitalize="sentences"
            disableFullscreenUI={true}
            editable={true}
            maxLength={255}
            multiline={false}
            onBlur={[Function]}
            onChange={[Function]}
            onChangeText={[Function]}
            onContentSizeChange={[Function]}
            onFocus={[Function]}
            secureTextEntry={false}
            selectionColor="#FFFFFF"
            style={
              [
                {
                  "flex": 1,
                  "includeFontPadding": false,
                  "margin": 0,
                  "padding": 0,
                  "paddingTop": 0,
                  "textAlign": "left",
                  "textAlignVertical": "top",
                  "top": 2,
                },
                {
                  "color": "#FFFFFF",
                  "fontSize": 16,
                  "height": 24,
                },
                [
                  {
                    "fontSize": 16,
                    "fontWeight": "normal",
                  },
                  {},
                ],
              ]
            }
            testID="text-input-default"
            underlineColorAndroid="transparent"
            value="Value TextInput"
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
          {
            "minHeight": 8,
            "paddingLeft": 0,
            "paddingRight": 0,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`TextInput component test TextInput snapshot test should render correctly 1`] = `
<View
  style={{}}
>
  <View
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="auto"
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "height": 52,
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "borderColor": "rgba(32, 32, 32, 1)",
              "borderStyle": "solid",
              "borderWidth": 1,
              "bottom": 0,
              "left": -1.5,
              "position": "absolute",
              "right": -1.5,
              "top": -1.5,
            }
          }
        />
      </View>
      <View
        pointerEvents="auto"
        style={
          {
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "left": "-100%",
              "paddingLeft": "50%",
              "position": "absolute",
              "top": 0,
              "transform": [
                {
                  "scale": 0.75,
                },
                {
                  "translateY": 0,
                },
                {
                  "translateX": 0,
                },
              ],
              "width": "200%",
            }
          }
        >
          <Text
            collapsable={false}
            numberOfLines={1}
            style={
              {
                "color": "rgba(32, 32, 32, 1)",
                "fontSize": 16,
                "fontWeight": "normal",
                "includeFontPadding": false,
                "lineHeight": 16,
                "textAlign": "left",
                "textAlignVertical": "top",
              }
            }
          >
            Label TextInput
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            accessibilityLabel="text-input-default"
            allowFontScaling={false}
            autoCapitalize="sentences"
            disableFullscreenUI={true}
            editable={true}
            maxLength={255}
            multiline={false}
            onBlur={[Function]}
            onChange={[Function]}
            onChangeText={[Function]}
            onContentSizeChange={[Function]}
            onFocus={[Function]}
            secureTextEntry={false}
            selectionColor="#ADADAD"
            style={
              [
                {
                  "flex": 1,
                  "includeFontPadding": false,
                  "margin": 0,
                  "padding": 0,
                  "paddingTop": 0,
                  "textAlign": "left",
                  "textAlignVertical": "top",
                  "top": 2,
                },
                {
                  "color": "#202020",
                  "fontSize": 16,
                  "height": 24,
                },
                [
                  {
                    "fontSize": 16,
                    "fontWeight": "normal",
                  },
                  {},
                ],
              ]
            }
            testID="text-input-default"
            underlineColorAndroid="transparent"
            value="Value TextInput"
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
          {
            "minHeight": 8,
            "paddingLeft": 0,
            "paddingRight": 0,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`TextInput component test TextInput with custom color snapshot test should render correctly 1`] = `
<View
  style={{}}
>
  <View
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="auto"
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "height": 52,
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "borderColor": "rgba(255, 255, 255, 1)",
              "borderStyle": "solid",
              "borderWidth": 1,
              "bottom": 0,
              "left": -1.5,
              "position": "absolute",
              "right": -1.5,
              "top": -1.5,
            }
          }
        />
      </View>
      <View
        pointerEvents="auto"
        style={
          {
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "left": "-100%",
              "paddingLeft": "50%",
              "position": "absolute",
              "top": 0,
              "transform": [
                {
                  "scale": 0.75,
                },
                {
                  "translateY": 0,
                },
                {
                  "translateX": 0,
                },
              ],
              "width": "200%",
            }
          }
        >
          <Text
            collapsable={false}
            numberOfLines={1}
            style={
              {
                "color": "rgba(255, 255, 255, 1)",
                "fontSize": 16,
                "fontWeight": "normal",
                "includeFontPadding": false,
                "lineHeight": 16,
                "textAlign": "left",
                "textAlignVertical": "top",
              }
            }
          >
            Label TextInput
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            accessibilityLabel="text-input-default"
            allowFontScaling={false}
            autoCapitalize="sentences"
            disableFullscreenUI={true}
            editable={true}
            maxLength={255}
            multiline={false}
            onBlur={[Function]}
            onChange={[Function]}
            onChangeText={[Function]}
            onContentSizeChange={[Function]}
            onFocus={[Function]}
            secureTextEntry={false}
            selectionColor="gray"
            style={
              [
                {
                  "flex": 1,
                  "includeFontPadding": false,
                  "margin": 0,
                  "padding": 0,
                  "paddingTop": 0,
                  "textAlign": "left",
                  "textAlignVertical": "top",
                  "top": 2,
                },
                {
                  "color": "black",
                  "fontSize": 16,
                  "height": 24,
                },
                [
                  {
                    "fontSize": 16,
                    "fontWeight": "normal",
                  },
                  {},
                ],
              ]
            }
            testID="text-input-default"
            underlineColorAndroid="transparent"
            value="Value TextInput"
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
          {
            "minHeight": 8,
            "paddingLeft": 0,
            "paddingRight": 0,
          },
        ]
      }
    />
  </View>
</View>
`;
